core/src/main/scala/org/apache/spark/deploy/security/HBaseDelegationTokenProvider.scala

```diff
@@ -50,12 +53,71 @@ private[security] class HBaseDelegationTokenProvider
     } catch {
       case NonFatal(e) =>
-        logWarning(s"Failed to get token from service $serviceName", e)
+        logWarning(s"Failed to get token from service $serviceName. " +
+          s"Retrying to invoke service with hbase connection.")
+        // Seems to be spark is trying to get the token from HBase 2.x.x  version or above where the
+        // obtainToken(Configuration conf) API is been removed. Lets try obtaining the token from
+        // another compatible API of HBase service.
+        obtainDelegationTokensWithHBaseConn(hadoopConf, creds)
     }
-
     None
   }
 
+  /**
+   * The HBase client API used in below method is introduced from HBase 0.98.9,1.0.0 version
+   * to invoke this api first connection object has to be retrieved from ConnectionFactory and the
+   * same connection can be passed to
+   * Token<AuthenticationTokenIdentifier> obtainToken(Connection conn) API
+   *
+   * @param hadoopConf
+   * @param creds
+   */
+  private def obtainDelegationTokensWithHBaseConn(
+      hadoopConf: Configuration, creds: Credentials): Unit = {
+    val mirror = universe.runtimeMirror(Utils.getContextOrSparkClassLoader)
+    // Token<AuthenticationTokenIdentifier> obtainToken(Configuration conf) is a deprecated
+    // method and in Hbase 2.0.0 the method is already removed. there is one more public
+    // consistent API Token<AuthenticationTokenIdentifier> obtainToken(Connection conn),
+    // in TokenUtil class which can be invoked using reflection.
+    val obtainHBaseConnection = mirror.classLoader.
+      loadClass("org.apache.hadoop.hbase.client.ConnectionFactory").
+      getMethod("createConnection", classOf[Configuration])
+    val conn = obtainHBaseConnection.invoke(null, hbaseConf(hadoopConf))
+    try {
+      val arrayOfMethod: Array[Method] = ClassLoader.getSystemClassLoader
+        .loadClass("org.apache.hadoop.hbase.security.token.TokenUtil").getMethods
+      val obtainParamTypeClass = getClassParamType(arrayOfMethod)
+      if (null == obtainParamTypeClass) {
+        logDebug(s"Failed to invoke service to get token from service $serviceName")
+      }
+      else {
+        val obtainTokenMethod = mirror.classLoader.
+          loadClass("org.apache.hadoop.hbase.security.token.TokenUtil").
+          getMethod("obtainToken", obtainParamTypeClass)
+        logDebug("Attempting to fetch HBase security token.")
+        val token = obtainTokenMethod.invoke(null, conn)
+          .asInstanceOf[Token[_ <: TokenIdentifier]]
+        logInfo(s"Get token from HBase: ${token.toString}")
+        creds.addToken(token.getService, token)
+      }
+    }
+    catch {
+      case NonFatal(e) =>
+        logError(s"Failed to get token from service $serviceName", e)
+    }
+    finally {
+      conn.asInstanceOf[Closeable].close()
+    }
+  }
+
+  private def getClassParamType(arrayOfMethod: Array[Method]): Class[_] = {
+    val seqOfFiltMethod = arrayOfMethod.filter(f => f.getName.equalsIgnoreCase("obtainToken"))
```

Comment:
Nits: just write `filter(_.getName....)`. You can't ignore case here, right?

Reply:
This Method is not required, as you told from reflection we can directly load the Class of the Connection  as below, we can omit the above logic.
  val connectionParamTypeClassRef = mirror.classLoader
        .loadClass("org.apache.hadoop.hbase.client.Connection")